File: 2024/day_04/solution.ts
(() => {
  function createXmasTree(height: number, ornament: string): string {
    let tree = "";

    for (let i = 1; i <= height; i++) {
      const spaces = height - i;
      const stars = i * 2 - 1;

      tree +=
        "_".repeat(spaces) + ornament.repeat(stars) + "_".repeat(spaces) + "\n";
    }

    tree += `${"_".repeat(height - 1) + "#" + "_".repeat(height - 1) + "\n"}`
      .repeat(2)
      .slice(0, -1);

    return tree;
  }

  console.log(createXmasTree(5, "*"), "\n");
  console.log(createXmasTree(3, "+"), "\n");
  console.log(createXmasTree(6, "@"), "\n");
})();

File: 2024/day_04/solution.py
def create_xmas_tree(height: int, ornament: str) -> str:
  tree = ""

  # Create tree rows
  for i in range(1, height + 1):
    spaces = height - i
    stars = i * 2 - 1
    tree += "_" * spaces + ornament * stars + "_" * spaces + "\n"

  # Create tree base
  base = ("_" * (height - 1) + "#" + "_" * (height - 1) + "\n") * 2

  # Delete last newline character
  tree += base.rstrip("\n")  

  return tree
